### PR TITLE
Peer trade wallet

### PR DESCRIPTION
**Pull request by alicevhenshaw**
- Fix for github issue https://github.com/airswap/airswap-protocols/issues/148
- Addition of more tests into Peer.js to increase the coverage of the integration tests to 100%

```
-----------------------|----------|----------|----------|----------|----------------|
File                   |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------|----------|----------|----------|----------|----------------|
 contracts/            |      100 |      100 |      100 |      100 |                |
  Imports.sol          |      100 |      100 |      100 |      100 |                |
  Peer.sol             |      100 |      100 |      100 |      100 |                |
 contracts/interfaces/ |      100 |      100 |      100 |      100 |                |
  IPeer.sol            |      100 |      100 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
All files              |      100 |      100 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
```

### CODE REVIEW

File: protocols/peer/contracts/Peer.sol [hunk 1]
```diff
@@ -215,6 +229,9 @@ contract Peer is IPeer, Ownable {
     require(_order.maker.wallet == msg.sender,
       "MAKER_MUST_BE_SENDER");
 
+    require(_order.taker.wallet == tradeWallet || _order.taker.wallet == address(0),
```
Comment (poiresel) [inline protocols/peer/contracts/Peer.sol]: how is it possible that _order.taker.wallet == address(0)?
if _order.taker.wallet is 0x0, then Swap.swap, sets the takerWallet to the Peer, however, the Peer doesn't have any funds nor can the owner withdraw tokens from it meaning the swap cannot succeed.

Comment (alicevhenshaw) [reply protocols/peer/contracts/Peer.sol]: Discussed in person and removed

File: protocols/peer-factory/test/Peer-factory-unit.js [hunk 2]
```diff
@@ -73,15 +80,20 @@ contract('Peer Factory Tests', async accounts => {
 
     it('should create delegate with the correct values', async () => {
```
Comment (poiresel) [inline protocols/peer-factory/test/Peer-factory-unit.js]: should create peer with correct values

File: protocols/peer-factory/test/Peer-factory-unit.js [hunk 3]
```diff
@@ -93,10 +105,16 @@ contract('Peer Factory Tests', async accounts => {
       let peer = await Peer.at(peerAddress)
       let actualSwap = await peer.swapContract.call()
       let actualOwner = await peer.owner.call()
+      let actualTradeWallet = await peer.tradeWallet.call()
 
       // check they are correct
```
Comment (poiresel) [inline protocols/peer-factory/test/Peer-factory-unit.js]: // check that the addresses are equal

File: protocols/peer/test/Peer.js [hunk 4]
```diff
@@ -3,17 +3,27 @@ const Swap = artifacts.require('Swap')
...
 
 contract('Peer', async accounts => {
-  let aliceAddress = accounts[0]
-  let bobAddress = accounts[1]
-  let carolAddress = accounts[2]
-  let davidAddress = accounts[3]
+  let aliceAddress = accounts[1]
+  let bobAddress = accounts[2]
+  let carolAddress = accounts[3]
+  let aliceTradeWallet = accounts[5]
```
Comment (poiresel) [inline protocols/peer/test/Peer.js]: any reason for skipping 4?

File: protocols/peer/test/Peer.js [hunk 5]
```diff
@@ -43,17 +48,40 @@ contract('Peer', async accounts => {
...
     await revertToSnapShot(snapshotId)
   })
 
-  describe('Deploying...', async () => {
-    it('Alice deployed a Swap Peer', async () => {
-      alicePeer = await Peer.new(swapAddress, aliceAddress, {
-        from: aliceAddress,
-      })
+  describe('Checks setTradeWallet', async () => {
+    it('Does not set a 0x0 trade wallet', async () => {
```
Comment (poiresel) [inline protocols/peer/test/Peer.js]: Cannot set a 0x0 trade wallet address

File: protocols/peer/test/Peer.js [hunk 6]
```diff
@@ -43,17 +48,40 @@ contract('Peer', async accounts => {
...
+      let val = await alicePeer.tradeWallet.call()
+      equal(val, carolAddress, 'trade wallet is incorrect')
+
+      //change it back
+      await alicePeer.setTradeWallet(aliceTradeWallet, { from: aliceAddress })
+    })
+
+    it('Non-owner cannot set a new address', async () => {
+      await reverted(
+        alicePeer.setTradeWallet(carolAddress, { from: carolAddress })
```
Comment (poiresel) [inline protocols/peer/test/Peer.js]: nice check!

File: protocols/peer/test/Peer.js (hunk 4 above)
Comment (alicevhenshaw) [reply protocols/peer/test/Peer.js]: oh david was after carol and then I realised he wasn't being used. Will update the trade wallet to 4.

File: protocols/peer/contracts/Peer.sol [hunk 7]
```diff
@@ -62,6 +76,7 @@ contract Peer is IPeer, Ownable {
     * @param _maxTakerAmount uint256 The maximum amount of ERC-20 token the peer would send
     * @param _priceCoef uint256 The whole number that will be multiplied by 10^(-priceExp) - the price coefficient
     * @param _priceExp uint256 The exponent of the price to indicate location of the decimal priceCoef * 10^(-priceExp)
+    * @dev 1 takerToken = priceCoef * 10^(-priceExp) makerToken
```
Comment (poiresel) [inline protocols/peer/contracts/Peer.sol]: move the @dev to be before the @params

**Comment by alicevhenshaw** [comment]
@poiresel that test already exists:
```
  it('Test unsetRule permissions', async () => {
      await reverted(
        peer.unsetRule(TAKER_TOKEN, MAKER_TOKEN, { from: notOwner })
      )
      await passes(peer.unsetRule(TAKER_TOKEN, MAKER_TOKEN, { from: owner }))
  })
```

File: protocols/peer/contracts/Peer.sol [hunk 8]
```diff
@@ -33,6 +33,9 @@ contract Peer is IPeer, Ownable {
   // Swap contract to be used to settle trades
   ISwap public swapContract;
 
+  // The address holding tokens that will be trading through this peer
+  address private tradeWalletAddress;
```
Comment (poiresel) [inline protocols/peer/contracts/Peer.sol]: make this _tradeWallet as it's a private variable and keep it consistent with its getter function

**Pull request merged**